### PR TITLE
Distribute reaction rates and densities when restarting under MPI

### LINKED ISSUE #1275: Restart depletion fails with MPI
When performing a restart run with multiple MPI processes, it appears that one operator tries building reaction rate data/material data before all the cross sections have been loaded.

## Steps to reproduce
1. Build a simple problem, and deplete for some period of time with some power
1. Create a new operator for restart mode by reading in from the previous `depletion_results.h5` file
1. Deplete again with some power and some time step but with multiple MPI processes.

I have two files: `start.py` and `restart.py` for the first depletion sequence and for the restart. Running `mpirun -n N restart.py` always fails, whether or not `start.py` used multiple processes.

## Error
`mpirun -n 2 restart.py 2>&1 > out.log`
```
 Reading c_H_in_H2O from /home/drew/nndc_hdf5/c_H_in_H2O.h5
 Maximum neutron transport energy: 20000000.000000 eV for U235
 Reading tallies XML file...
 Writing summary.h5 file...
Time to matexp:  0.1556260585784912
Traceback (most recent call last):
  File "restart.py", line 20, in <module>
    openmc.deplete.integrator.predictor(op, [time_steps[0]], power)
  File "/home/drew/openmc/openmc/deplete/integrator/predictor.py", line 100, in predictor
    op_results = [operator(x[0], power[-1])]
  File "/home/drew/openmc/openmc/deplete/operator.py", line 175, in __call__
    self.number.set_density(vec)
  File "/home/drew/openmc/openmc/deplete/atom_number.py", line 214, in set_density
    self.set_mat_slice(i, density_slice)
  File "/home/drew/openmc/openmc/deplete/atom_number.py", line 199, in set_mat_slice
    self[mat, :self.n_nuc_burn] = val
  File "/home/drew/openmc/openmc/deplete/atom_number.py", line 104, in __setitem__
    self.number[mat, nuc] = val
IndexError: index 0 is out of bounds for axis 0 with size 0
 Reading I135 from /home/drew/nndc_hdf5/I135.h5
 Reading Xe135 from /home/drew/nndc_hdf5/Xe135.h5
```
[Full error - pastebin](https://pastebin.com/RwaDfNc5)

The process also hung at the last line

### PR DESCRIPTION
**Pull request by drewejohnson**
All the comments here are related to restarting with multiple MPI processes. Otherwise this bug doesn't exists.

This PR closes #1275 by creating a specific `ResultsList` for each `Operator` / MPI process. The main result file contains reaction rates and numbers for all burnable materials, while each `Operator` is only responsible for depleting a subset of materials. The `ResultsList` created contains views into the main `data` [numbers] and `rates` arrays from the complete `ResultsList`. This is done with a new method: `Results.distribute` which is a fancier `copy` method for creating a new `Results` object with views corresponding to process-specific materials.

I compared these by depleting a quarter PWR assembly with 72 uniquely burnable materials. Both cases started from the same restart file, with one case running in serial and the other using two MPI processes. Comparisons were performed with `h5diff` and showed no changes over 0.01 percent between number densities, eigenvalues, and reaction rates.

### CODE REVIEW

File: openmc/deplete/results.py [hunk 1]
```diff
@@ -153,6 +153,34 @@ def allocate(self, volume, nuc_list, burn_list, full_burn_list, stages):
...
+        Parameters
+        ----------
+        local_materials : iterable of str
+            Materials for this process
+        ranges : iterable of int
+
+        Returns
+        -------
+        Results
+            - New results object
```
Comment (paulromano) [inline openmc/deplete/results.py]:
```suggestion
            New results object
```

File: openmc/deplete/results.py [hunk 2]
```diff
@@ -153,6 +153,34 @@ def allocate(self, volume, nuc_list, burn_list, full_burn_list, stages):
...
         self.data = np.zeros((stages, self.n_mat, self.n_nuc))
 
+    def distribute(self, local_materials, ranges):
+        """Create a new object containing data for distributed materials
+
+        Parameters
+        ----------
+        local_materials : iterable of str
+            Materials for this process
+        ranges : iterable of int
```
Comment (paulromano) [inline openmc/deplete/results.py]: Need description for this

File: openmc/deplete/results_list.py [hunk 3]
```diff
@@ -8,22 +8,34 @@
...
-        The filename to read from.
-
+    It is recommended to use :meth:`from_hdf5` over
+    direct creation.
     """
-    def __init__(self, filename):
-        super().__init__()
+
+    @classmethod
+    def from_hdf5(cls, filename):
```
Comment (paulromano) [inline openmc/deplete/results_list.py]: Why not just keep this as `__init__`?

File: openmc/deplete/results_list.py [hunk 4]
```diff
@@ -8,22 +8,34 @@
...
+
+        Parameters
+        ----------
+        filename : str
+            Path to depletion result file
+
+        Returns
+        -------
+        new : ResultsList
+           New instance of depletion results
```
Comment (paulromano) [inline openmc/deplete/results_list.py]:
```suggestion
            New instance of depletion results
```

File: openmc/deplete/results_list.py (hunk 3 above)
Comment (paulromano) [reply openmc/deplete/results_list.py]: Nevermind, I see that it is needed for the distribute logic

**Comment by drewejohnson** [comment]
I would prefer to merge #1287 first, because e286940f8eb3e86ac318fdfe146eec6e23deed9a passes the `prev_results` onto the `TransportOperator` base class. For this PR, we could add a new abstract method like `distribute_prev_results` where the `Operator` would perform this distribution. Not sure if that method should return the distributed results or set them directly, e.g.
```
self.prev_results = self.distribute_prev_results(prev_results)
```
vs
```
self.distribute_prev_results(prev_results)
```

**Pull request merged**